File: tests/e2e/pages/CartPage.ts
import { Page, expect } from '@playwright/test';

export class CartPage {
    constructor(private readonly page: Page) {}

    async navigate() {
        await this.page.goto('/customer/cart');
        await expect(this.page).toHaveURL(/\/customer\/cart/);
    }

    async applyPromo(code: string) {
        const promoInput = this.page.locator('input[name="promo_code"]');
        const applyButton = this.page.getByRole('button', { name: /apply/i });

        await expect(promoInput).toBeVisible();
        await promoInput.fill(code);
        await expect(applyButton).toBeVisible();
        await applyButton.click();
    }

    async verifyDiscountApplied() {
        const successMsg = this.page.locator('.promo-msg');
        await expect(successMsg).toBeVisible();
        await expect(successMsg).toContainText(/code applied successfully/i);
    }
}

File: tests/E2E/pages/cart_page.py
class CartPage:
    def __init__(self, client):
        self.client = client

    def add_item(self, item_id, quantity=1):
        return self.client.post("/customer/api/cart/add", json={
            "item_id": item_id,
            "quantity": quantity,
        })

    def apply_promo(self, code):
        return self.client.post("/customer/cart", data={
            "promo_code": code,
        })